pgxn/neon/walproposer_pg.c

```diff
@@ -1482,6 +1482,21 @@ walprop_pg_wait_event_set(WalProposer *wp, long timeout, Safekeeper **sk, uint32
 #if PG_MAJORVERSION_NUM >= 16
 	if (WalSndCtl != NULL)
 		ConditionVariablePrepareToSleep(&WalSndCtl->wal_flush_cv);
+
+	/*
+	 * Now that we prepared the condvar, check flush ptr again -- it might have
+	 * changed before we subscribed to cv so we missed the wakeup.
+	 *
+	 * Do that only when we're interested in new WAL: without sync-safekeepers
+	 * and if election already passed.
+	 */
+	if (!wp->config->syncSafekeepers && wp->availableLsn != InvalidXLogRecPtr && GetFlushRecPtr(NULL) > wp->availableLsn)
```

Comment:
I wonder what is the difference between our walproposer and standard Postgres walsender?
Why we have to do this extra recheck after preparing condor but before waiting and vanilla Postgres is not doing it?
There is no slowdown in sync replication in pg16 - the same example works fast with vanilla pg16 and synchronous replica.

Reply:
Because vanilla ignores WaitEventSetWait result; it tries to send WAL even if walsender has been waken up by socket event, not latch set. So replica reply wakes walsender and makes it send WAL. While this is generally incorrect, it is really hard to make up a scenario where delay is noticeable: with sync replication backend can't proceed until reply comes anyway, with async replication backends are not blocked at all and continue wake up walsenders on flushes. It probably could make async replication latency higher in really unlucky rare schedules, but again reply "fixes" this, so latency just becomes 2x at worst. Which is hard to notice if it is small (it's async after all) and doesn't much matter if it is large.

In walproposer, we try to send WAL only if latch has been set.